File: examples/enterprise/rust/src/bin/ent_10_s3_etl_pipeline.rs
//! Enterprise 10 · S3 ETL pipeline — embedded mode (LocalFileSystem)
//!
//! Demonstrates `S3Source` + `S3Sink` with `object_store::local::LocalFileSystem`
//! as a drop-in for real S3/MinIO:
//!
//!   1. Write synthetic IoT sensor data via ArrowWriter (simulating an upload)
//!   2. Open it back with `S3Source` (streaming Parquet reader)
//!   3. Register batches in Krishiv, run SQL transform (filter + enrich)
//!   4. Write the curated result to a new location via `S3Sink`
//!
//! In production, swap `LocalFileSystem::new(dir)` for an `AmazonS3` or
//! `MicrosoftAzure` object store — the connector code is identical.
//!
//! Run:
//!   cargo run -p krishiv-enterprise-examples --bin ent_10_s3_etl_pipeline

use std::sync::Arc;

use anyhow::{Context, Result};
use arrow::array::{Float64Array, Int64Array, StringArray};
use arrow::datatypes::{DataType, Field, Schema};
use arrow::record_batch::RecordBatch;
use krishiv::Session;
use krishiv_connectors::s3::{S3Sink, S3Source};
use krishiv_connectors::Source;
use object_store::{ObjectStoreExt as _, PutPayload, local::LocalFileSystem, path::Path as StorePath};
use parquet::arrow::ArrowWriter;
use tempfile::tempdir;

#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Enterprise 10: S3 ETL Pipeline (embedded / LocalFileSystem) ===");

    let dir = tempdir()?;
    let store: Arc<dyn object_store::ObjectStore> =
        Arc::new(LocalFileSystem::new_with_prefix(dir.path())?);

    let input_path  = StorePath::from("input/sensors.parquet");
    let output_path = StorePath::from("output/curated.parquet");

    // ── Step 1: write synthetic sensor data (simulates an S3 upload) ──────
    println!("  writing synthetic sensors to {}", input_path);
    let sensor_schema = Arc::new(Schema::new(vec![
        Field::new("sensor_id",     DataType::Utf8,    false),
        Field::new("ts_ms",         DataType::Int64,   false),
        Field::new("temp_c",        DataType::Float64, false),
        Field::new("humidity_pct",  DataType::Float64, false),
        Field::new("location",      DataType::Utf8,    false),
    ]));

    let base: i64 = 1_716_200_000_000;
    let sensor_batch = RecordBatch::try_new(sensor_schema.clone(), vec![
        Arc::new(StringArray::from(vec![
            "s001","s001","s001","s001",
            "s002","s002","s002","s002",
            "s003","s003","s003",
        ])),
        Arc::new(Int64Array::from(vec![
            base, base+60_000, base+120_000, base+180_000,
            base+10_000, base+70_000, base+130_000, base+190_000,
            base+20_000, base+80_000, base+140_000,
        ])),
        Arc::new(Float64Array::from(vec![
            22.1, 22.5, 23.0, 22.8,       // s001 — comfortable
            -3.5, -2.0, -4.1, 39.5,       // s002 — cold-storage (39.5 = anomaly)
            18.0, 18.2, 18.5,             // s003 — office
        ])),
        Arc::new(Float64Array::from(vec![
            45.0, 44.5, 46.0, 45.5,
            80.0, 78.0, 82.0, 101.0,     // 101% = invalid humidity
            55.0, 54.0, 56.0,
        ])),
        Arc::new(StringArray::from(vec![
            "warehouse-a","warehouse-a","warehouse-a","warehouse-a",
            "cold-storage","cold-storage","cold-storage","cold-storage",
            "office","office","office",
        ])),
    ])?;

    // Write Parquet bytes to a temp file, then upload to the object store.
    let tmp_input = dir.path().join("_sensors_tmp.parquet");
    {
        let file = std::fs::File::create(&tmp_input)?;
        let mut w = ArrowWriter::try_new(file, sensor_schema.clone(), None)?;
        w.write(&sensor_batch)?;
        w.close()?;
    }
    let raw = std::fs::read(&tmp_input)?;
    store.put(&input_path, PutPayload::from(raw)).await.context("put input")?;

    // ── Step 2: read back via S3Source ────────────────────────────────────
    println!("  reading via S3Source from {}", input_path);
    let mut source = S3Source::open(Arc::clone(&store), input_path.clone())
        .await
        .context("S3Source::open")?;

    let mut raw_batches: Vec<RecordBatch> = Vec::new();
    while let Some(batch) = source.read_batch().await.context("read_batch")? {
        raw_batches.push(batch);
    }
    let raw_rows: usize = raw_batches.iter().map(|b| b.num_rows()).sum();
    println!("  read {} raw rows", raw_rows);

    // ── Step 3: SQL transform ─────────────────────────────────────────────
    let session = Session::builder().build()?;
    session.register_record_batches("raw_sensors", raw_batches)?;

    let df = session.sql(
        "SELECT
             sensor_id,
             ts_ms,
             temp_c,
             CAST(temp_c * 9.0 / 5.0 + 32.0 AS DOUBLE) AS temp_f,
             humidity_pct,
             location,
             CASE
                 WHEN temp_c < 0  THEN 'cold'
                 WHEN temp_c < 25 THEN 'comfortable'
                 ELSE 'hot'
             END AS temp_bucket,
             CAST((temp_c > 38 OR temp_c < -5) AS BOOLEAN) AS is_anomaly
         FROM raw_sensors
         WHERE humidity_pct BETWEEN 10 AND 95
         ORDER BY ts_ms",
    )?;

    let curated_result = df.collect()?;
    let curated_batches = curated_result.into_batches();
    let curated_rows: usize = curated_batches.iter().map(|b| b.num_rows()).sum();
    println!("  {} curated rows after filter (removed invalid humidity)", curated_rows);

    // ── Step 4: write to S3Sink ───────────────────────────────────────────
    let mut sink = S3Sink::new(Arc::clone(&store), output_path.clone());

    use krishiv_connectors::Sink;
    for batch in &curated_batches {
        sink.write_batch(batch.clone()).await.context("write_batch")?;
    }
    sink.flush().await.context("flush")?;

    println!("  curated parquet written to {}", output_path);

    // ── Verify: read curated back via S3Source ────────────────────────────
    let mut read_back = S3Source::open(Arc::clone(&store), output_path)
        .await
        .context("open curated")?;
    let mut verify: Vec<RecordBatch> = Vec::new();
    while let Some(b) = read_back.read_batch().await? {
        verify.push(b);
    }

    let verify_rows: usize = verify.iter().map(|b| b.num_rows()).sum();
    println!("\n✓ ETL complete  raw={} → curated={} rows", raw_rows, verify_rows);

    let session2 = Session::builder().build()?;
    session2.register_record_batches("curated", verify)?;
    let summary = session2.sql(
        "SELECT temp_bucket, COUNT(*) AS count, ROUND(AVG(temp_c), 2) AS avg_temp, \
                SUM(CASE WHEN is_anomaly THEN 1 ELSE 0 END) AS anomalies \
         FROM curated GROUP BY temp_bucket ORDER BY avg_temp"
    )?;
    println!("\n--- Curated summary by temp bucket ---");
    println!("{}", summary.collect()?.pretty()?);

    Ok(())
}

File: examples/enterprise/python/ent_10_s3_etl_pipeline.py
"""Enterprise 10 · S3/MinIO ETL pipeline: read Parquet → transform → write output

Full batch ETL using S3-compatible storage (LocalStack or MinIO):
  1. Generate synthetic IoT sensor data and upload to s3://krishiv-data/input/
  2. Download, filter, and enrich with pandas SQL-style operations
  3. Write curated output to s3://krishiv-data/output/curated.parquet

In production, remove AWS_ENDPOINT_URL to target real AWS S3.

Prerequisites:
    make infra-up seed-aws   (creates the bucket)

Run:
    AWS_ENDPOINT_URL=http://localhost:4566 \
    AWS_ACCESS_KEY_ID=krishiv AWS_SECRET_ACCESS_KEY=krishiv \
    python python/ent_10_s3_etl_pipeline.py
"""

import io
import os

import boto3
import pyarrow as pa
import pyarrow.parquet as pq
import pandas as pd

S3_BUCKET  = "krishiv-data"
INPUT_KEY  = "input/sensors.parquet"
OUTPUT_KEY = "output/curated.parquet"


def get_s3_client():
    kwargs = {
        "region_name":          "us-east-1",
        "aws_access_key_id":    os.environ.get("AWS_ACCESS_KEY_ID",     "krishiv"),
        "aws_secret_access_key": os.environ.get("AWS_SECRET_ACCESS_KEY", "krishiv"),
    }
    endpoint = os.environ.get("AWS_ENDPOINT_URL")
    if endpoint:
        kwargs["endpoint_url"] = endpoint
    return boto3.client("s3", **kwargs)


def main() -> None:
    endpoint = os.environ.get("AWS_ENDPOINT_URL", "https://s3.amazonaws.com")
    print("=== Enterprise 10 (Python): S3 ETL Pipeline ===")
    print(f"  S3 endpoint : {endpoint}")
    print(f"  input       : s3://{S3_BUCKET}/{INPUT_KEY}")
    print(f"  output      : s3://{S3_BUCKET}/{OUTPUT_KEY}")

    s3 = get_s3_client()

    # 1 ── Ensure bucket exists (LocalStack / MinIO).
    try:
        s3.head_bucket(Bucket=S3_BUCKET)
    except Exception:
        s3.create_bucket(Bucket=S3_BUCKET)
        print(f"  created bucket s3://{S3_BUCKET}")

    # 2 ── Upload synthetic input data if not already present.
    try:
        s3.head_object(Bucket=S3_BUCKET, Key=INPUT_KEY)
        print("  input already in S3 — downloading…")
        raw = _s3_download(s3, INPUT_KEY)
        df = pq.read_table(io.BytesIO(raw)).to_pandas()
    except Exception:
        print("  generating synthetic sensor data…")
        df = _make_sensor_df()
        buf = io.BytesIO()
        pq.write_table(pa.Table.from_pandas(df), buf)
        buf.seek(0)
        s3.put_object(Bucket=S3_BUCKET, Key=INPUT_KEY, Body=buf.read())
        print(f"  uploaded {len(df)} sensor records to s3://{S3_BUCKET}/{INPUT_KEY}")

    # 3 ── SQL-style transformation with pandas.
    print(f"\n  raw sensor stats ({len(df)} records):")
    print(df.groupby("sensor_id")["temp_c"].agg(["count", "mean", "max"]).round(2))

    # Filter invalid humidity, add Fahrenheit, bucket by temperature.
    curated = df[df["humidity_pct"].between(10, 95)].copy()
    curated["temp_f"] = (curated["temp_c"] * 9 / 5 + 32).round(2)
    curated["temp_bucket"] = pd.cut(
        curated["temp_c"],
        bins=[-float("inf"), 0, 15, 25, 35, float("inf")],
        labels=["freezing", "cold", "comfortable", "warm", "hot"],
    ).astype(str)
    curated["is_anomaly"] = (curated["temp_c"] > 38) | (curated["temp_c"] < -5)

    print(f"\n  {len(curated)} records after filter (removed {len(df) - len(curated)} invalid)")

    # 4 ── Upload curated output to S3.
    buf = io.BytesIO()
    pq.write_table(pa.Table.from_pandas(curated), buf)
    buf.seek(0)
    s3.put_object(Bucket=S3_BUCKET, Key=OUTPUT_KEY, Body=buf.read())
    print(f"\n✓ curated {len(curated)} rows → s3://{S3_BUCKET}/{OUTPUT_KEY}")

    # 5 ── Summary.
    summary = curated.groupby("temp_bucket").agg(
        count=("sensor_id", "count"),
        avg_temp=("temp_c", "mean"),
        anomalies=("is_anomaly", "sum"),
    ).round(2)
    print("\n--- Curated summary by temp bucket ---")
    print(summary.to_string())


def _s3_download(s3, key: str) -> bytes:
    obj = s3.get_object(Bucket=S3_BUCKET, Key=key)
    return obj["Body"].read()


def _make_sensor_df() -> pd.DataFrame:
    base = 1_716_200_000_000
    return pd.DataFrame([
        {"sensor_id": "s001", "ts_ms": base + 0,       "temp_c": 22.1, "humidity_pct": 45.0, "location": "warehouse-a"},
        {"sensor_id": "s001", "ts_ms": base + 60_000,  "temp_c": 22.5, "humidity_pct": 44.5, "location": "warehouse-a"},
        {"sensor_id": "s001", "ts_ms": base + 120_000, "temp_c": 23.0, "humidity_pct": 46.0, "location": "warehouse-a"},
        {"sensor_id": "s001", "ts_ms": base + 180_000, "temp_c": 22.8, "humidity_pct": 45.5, "location": "warehouse-a"},
        {"sensor_id": "s002", "ts_ms": base + 10_000,  "temp_c": -3.5, "humidity_pct": 80.0, "location": "cold-storage"},
        {"sensor_id": "s002", "ts_ms": base + 70_000,  "temp_c": -2.0, "humidity_pct": 78.0, "location": "cold-storage"},
        {"sensor_id": "s002", "ts_ms": base + 130_000, "temp_c": -4.1, "humidity_pct": 82.0, "location": "cold-storage"},
        {"sensor_id": "s002", "ts_ms": base + 190_000, "temp_c": 39.5, "humidity_pct": 101.0,"location": "cold-storage"},  # invalid humidity
        {"sensor_id": "s003", "ts_ms": base + 20_000,  "temp_c": 18.0, "humidity_pct": 55.0, "location": "office"},
        {"sensor_id": "s003", "ts_ms": base + 80_000,  "temp_c": 18.2, "humidity_pct": 54.0, "location": "office"},
        {"sensor_id": "s003", "ts_ms": base + 140_000, "temp_c": 18.5, "humidity_pct": 56.0, "location": "office"},
    ])


if __name__ == "__main__":
    main()
